### PR TITLE
Added Change Feed

### PR DESCRIPTION
**Pull request by seanmcc-msft**

### CODE REVIEW

**Comment by seanmcc-msft** [comment]
TODO - Add more unit tests.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Shard.cs [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+            return changeFeedEvent;
+        }
+
+        /// <inheritdoc/>
+        public void Dispose() => _currentChunk.Dispose();
+
+        /// <summary>
+        /// Constructor for testing.  Do not use.
+        /// </summary>
+        internal Shard(
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Shard.cs:165]: These test-only constructors weird me out, but because these object are stateful, it's challenging to get Shard, Chunk, Segment, and Change Feed into the correct starting state for the smaller unit tests.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Models/BlobChangeFeedEvent.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+using System.Diagnostics;
+using System.Globalization;
+using System.Text;
+
+namespace Azure.Storage.Blobs.ChangeFeed.Models
+{
+    /// <summary>
+    /// BlobChangeFeedEvent.
+    /// </summary>
+    public class BlobChangeFeedEvent
```
Comment (KrzysztofCwalina) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Models/BlobChangeFeedEvent.cs:15]: I would call it BlobChangeEvent

**Comment by seanmcc-msft** [comment]
Here's the API view (slightly out of date) - https://apiview.dev/Assemblies/Review/85046278a8764a02a11d65ab219b91ae

@tg-msft has recommendations for the public model names.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/tests/SegmentTests.cs [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        }
+
+        /// <summary>
+        /// In this test, the Segment has 3 Shards and 5 total Events.
+        /// Shard index 0 and 1 have 2 Events, and Shard index 2 has 1 Event.
+        /// We are round-robining the Shards, so we will return the events for
+        /// the shards indexes: 0 1 2 0 1.
+        /// </summary>
+        [Test]
+        public async Task GetPage()
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/tests/SegmentTests.cs:85]: @gapra-msft, here's my Segment round robin test.

**Comment by seanmcc-msft** [comment]
@kasobol-msft, @tg-msft, @gapra-msft, could you take a look?

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+        /// <param name="containerClient"></param>
+        internal ChangeFeed(
+            BlobContainerClient containerClient)
+        {
+            _containerClient = containerClient;
+        }
+
+        private async Task Initalize(bool async)
+        {
+            // Check if Change Feed has been abled for this account.
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs:106]: typo

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+            }
+            else
+            {
+                changeFeedContainerExists = _containerClient.Exists();
+            }
+
+            if (!changeFeedContainerExists)
+            {
+                //TODO improve this error message
+                throw new ArgumentException("Change Feed hasn't been enabled on this account, or is current being enabled.");
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs:121]: current->currently

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+            if (async)
+            {
+                jsonMetaSegment = await JsonDocument.ParseAsync(blobDownloadInfo.Content).ConfigureAwait(false);
+            }
+            else
+            {
+                jsonMetaSegment = JsonDocument.Parse(blobDownloadInfo.Content);
+            }
+
+            //TODO what happens when _lastConsumable advances an hour?
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs:146]: Maybe we should just default end time to "now" when change feed is requested ?

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+            {
+                return new BlobChangeFeedEventPage();
+            }
+
+            //TODO what should we return here?  Also do we really need to check this on every page?
+            if (_currentSegment.DateTime > _lastConsumable)
+            {
+                return new BlobChangeFeedEventPage();
+            }
+
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs:212]: Shouldn't this be end of iterable?

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Models/BlobChangeFeedEventPage.cs [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+        public BlobChangeFeedEventPage() { }
+
+        public BlobChangeFeedEventPage(List<BlobChangeFeedEvent> events, string continuationToken)
+        {
+            Values = events;
+            ContinuationToken = continuationToken;
+        }
+
+    //    public BlobChangeFeedEventPage(Response raw, List<GenericRecord> data)
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Models/BlobChangeFeedEventPage.cs:25]: commented out code.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Segment.cs [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+            }
+
+            //TODO how to get raw response for page?  Does it matter?
+            return changeFeedEventList;
+        }
+
+        //TODO figure out if this is right.
+        public bool HasNext()
+        {
+            if (!_isInitalized)
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Segment.cs:172]: Should this trigger initialize instead? or throw? Good test for this assumption might be - fetch change log for account with zero activity.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Shard.cs (hunk 1 above)
Comment (kasobol-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Shard.cs:168]: I'd rather mock BlobContainerClient. Over time this might get quite out of sync from "prod" code leading to situation where we're not testing real scenarios.

**Review by kasobol-msft — COMMENTED** [review]
Overall looks good.
- few comments left
- quite a lot of TODOs seem to be unresolved
- Sample tests / readmes are empty.
- quite a lot of ignored tests.

**Comment by seanmcc-msft** [comment]
> I'd rather mock BlobContainerClient. Over time this might get quite out of sync from "prod" code leading to situation where we're not testing real scenarios.

I am blocking the BlobContainerClient, I have that constructor to avoid super complicate test setups to mock the functionality of Shard.Initalize().

**Comment by kasobol-msft** [comment]
> > I'd rather mock BlobContainerClient. Over time this might get quite out of sync from "prod" code leading to situation where we're not testing real scenarios.
> 
> I am mocking the BlobContainerClient, I have that constructor to avoid super complicate test setups to mock the functionality of Shard.Initalize().

Maybe `Initialize` should be converted into static factory method that returns fully initialized Shard then ?

https://www.reddit.com/r/learnprogramming/comments/a8j4cg/requiring_a_second_initialization_step_is_an/

**Comment by seanmcc-msft** [comment]
> > > I'd rather mock BlobContainerClient. Over time this might get quite out of sync from "prod" code leading to situation where we're not testing real scenarios.
> > 
> > 
> > I am mocking the BlobContainerClient, I have that constructor to avoid super complicate test setups to mock the functionality of Shard.Initalize().
> 
> Maybe `Initialize` should be converted into static factory method that returns fully initialized Shard then ?
> 
> https://www.reddit.com/r/learnprogramming/comments/a8j4cg/requiring_a_second_initialization_step_is_an/

Would there be any concurrency or perf issues if we switched to static factory methods?  Several Shards could be being initialized at the same time.

Just for background, the Initialize() methods are separate from the constructors because I don't think we can make async calls in the constructors.

**Comment by seanmcc-msft** [comment]
Another issue is I'm not sure if it's possible to mock static methods - https://stackoverflow.com/questions/12580015/how-to-mock-static-methods-in-c-sharp-using-moq-framework

**Comment by kasobol-msft** [comment]
> > > > I'd rather mock BlobContainerClient. Over time this might get quite out of sync from "prod" code leading to situation where we're not testing real scenarios.
> > > 
> > > 
> > > I am mocking the BlobContainerClient, I have that constructor to avoid super complicate test setups to mock the functionality of Shard.Initalize().
> > 
> > 
> > Maybe `Initialize` should be converted into static factory method that returns fully initialized Shard then ?
> > https://www.reddit.com/r/learnprogramming/comments/a8j4cg/requiring_a_second_initialization_step_is_an/
> 
> Would there be any concurrency or perf issues if we switched to static factory methods? Several Shards could be being initialized at the same time.
> 
> Just for background, the Initialize() methods are separate from the constructors because I don't think we can make async calls in the constructors.

Yeah. Lack of async constructors is known problem to deal with. On the other hand need for async constructor/initialization suggests that it's a heavy process and probably deserves separation (i.e. another class/method that does heavy lifting to a factory or static factory method or builder - depending on situation)

Perf shouldn't be a problem. Concurrency - factory method/class should be stateless.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Models/BlobChangeFeedEvent.cs (hunk 2 above)
Comment (tg-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Models/BlobChangeFeedEvent.cs:15]: I think we should consider leaving this as `BlobChangeFeedEvent` because we've got the same schema for changes routed through EventGrid and we don't have a Track 2 API or name for server-side logging event info yet.  `BlobChangeFeedEvent` is a little verbose, but I don't want to take a good name that we might want for the main Blobs package someday.

**Comment by seanmcc-msft** [comment]
I'm not sure the Factories would be viable - there's no good way to verify the state of the ChangeFeed/Segment/Shard/Chunk after it's created by the Factory.

**Comment by seanmcc-msft** [comment]
@kasobol-msft, I created Factory classes and much better mocked tests.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Shard.cs (hunk 1 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Shard.cs]: Fixed!

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs (hunk 6 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs]: I think we just default to the last hour present in the Change Feed.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs (hunk 4 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs]: fixed.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs (hunk 5 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs]: Fixed.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Models/BlobChangeFeedEventPage.cs (hunk 8 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Models/BlobChangeFeedEventPage.cs:25]: Fixed.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Segment.cs (hunk 9 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Segment.cs]: Now we that have factories, _isInitalized isn't a thing anymore in ChangeFeed, Segment, or Shard.

**Comment by seanmcc-msft** [comment]
@kasobol-msft, I filled in the readme and added code samples.

**Review by kasobol-msft — APPROVED** [review]

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/README.md [hunk 10]
```diff
@@ -0,0 +1,143 @@
+# Azure Storage Blobs Change Feed client library for .NET
```
Comment (KrzysztofCwalina) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/README.md:1]: Very nice readme. Thanks for writing it!

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/BlobChangeFeedClient.cs [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+using System.Threading.Tasks;
+using Azure.Storage.Blobs;
+using Azure.Storage.Blobs.ChangeFeed.Models;
+
+namespace Azure.Storage.Blobs.ChangeFeed
+{
+    /// <summary>
+    /// BlobChangeFeedClient.
+    /// </summary>
+    public class BlobChangeFeedClient
```
Comment (KrzysztofCwalina) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/BlobChangeFeedClient.cs:16]: Types with client suffix need to have public ctors. i.e. "Client" means a type that users can start with.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/BlobChangeFeedExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+        /// Builds a DateTimeOffset from a segment path.
+        /// </summary>
+        internal static DateTimeOffset? ToDateTimeOffset(this string segmentPath)
+        {
+            if (segmentPath == null)
+            {
+                return default;
+            }
+            segmentPath = segmentPath.Trim('/');
+            string[] splitPath = segmentPath.Split('/');
```
Comment (KrzysztofCwalina) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/BlobChangeFeedExtensions.cs:38]: I think if you do segmentPath.Split('/', StringSplitOptions.RemoveEmptyEntries, you don't have to trim first.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/BlobChangeFeedExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+            return new DateTimeOffset(
+                year: int.Parse(splitPath[2], CultureInfo.InvariantCulture),
```
Comment (KrzysztofCwalina) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/BlobChangeFeedExtensions.cs:40]: I think we should guard against index out of range.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/BlobChangeFeedClient.cs (hunk 11 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/BlobChangeFeedClient.cs:16]: We could add public constructors.  Another option would be to make `GetChanges()` an extension method to `BlobServiceClient`, but we won't be able to do the same for Java and Python.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/BlobChangeFeedExtensions.cs (hunk 12 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/BlobChangeFeedExtensions.cs:38]: Fixed.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/BlobChangeFeedExtensions.cs (hunk 13 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/BlobChangeFeedExtensions.cs:40]: Fixed.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/BlobChangeFeedClient.cs (hunk 11 above)
Comment (KrzysztofCwalina) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/BlobChangeFeedClient.cs:17]: The two options are: a) a public ctor, b) rename the type such that it does not have a "Client" suffix.

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/BlobChangeFeedClient.cs:17]: I opted to add the public constructors, I figure there's no disadvantage to allowing customers to new up a `BlobChangeFeedClient` directly.

**Comment by seanmcc-msft** [comment]
@tg-msft, could you take a look when you have time?

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeedFactory.cs [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+                changeFeedContainerExists = await containerClient.ExistsAsync().ConfigureAwait(false);
+            }
+            else
+            {
+                changeFeedContainerExists = containerClient.Exists();
+            }
+
+            if (!changeFeedContainerExists)
+            {
+                throw new ArgumentException("Change Feed hasn't been enabled on this account, or is currently being enabled.");
```
Comment (ljian3377) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeedFactory.cs:74]: or is currently being **disabled**

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeedFactory.cs [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+                containerClient: containerClient,
+                yearPath: firstYearPath,
+                startTime: startTime,
+                endTime: BlobChangeFeedExtensions.MinDateTime(lastConsumable, endTime))
+                .ConfigureAwait(false);
+
+            Segment currentSegment = await _segmentFactory.BuildSegment(
+                async,
+                containerClient,
+                segments.Dequeue(),
```
Comment (ljian3377) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeedFactory.cs:133]: The `segments[]` could be empty if there is no segment in that year after the `startTime`. I added a while loop to fetch the segments till it's not empty.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/BlobChangeFeedExtensions.cs [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+        internal static DateTimeOffset? RoundUpToNearestHour(this DateTimeOffset? dateTimeOffset)
+        {
+            if (dateTimeOffset == null)
+            {
+                return null;
+            }
+
+            DateTimeOffset? newDateTimeOffest = dateTimeOffset.RoundDownToNearestHour();
+
+            return newDateTimeOffest.Value.AddHours(1);
```
Comment (ljian3377) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/BlobChangeFeedExtensions.cs:92]: Is this right? If input is 15:00:00, it will return 16:00:00?

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/BlobChangeFeedExtensions.cs:92]: Correct.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+                throw new InvalidOperationException("Change feed doesn't have any more events");
+            }
+
+            if (_currentSegment.DateTime > _endTime)
+            {
+                return new BlobChangeFeedEventPage();
+            }
+
+            if (_currentSegment.DateTime > _lastConsumable)
+            {
```
Comment (ljian3377) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs:100]: Is this enough for `lastConsumable` checking?
I found the lastConsumable is not rounded to hour in the sample meta.json.
```json
{
    "version": 0,
    "lastConsumable": "2019-02-23T01:10:00.000Z",
    "storageDiagnostics": {
        "version": 0,
        "lastModifiedTime": "2019-02-23T02:24:00.556Z",
        "data": {
            "aid": "55e551e3-8006-0000-00da-ca346706bfe4",
            "lfz": "2019-02-22T19:10:00.000Z"
        }
    }
}
```
So if segment's  DateTime is 2019-02-23T01:00:00, then we may end up returning events later then lastConsumable?

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Segment.cs [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+                changeFeedEventList.Add(changeFeedEvent);
+
+                // If the current shard is completed, remove it from _shards
+                if (!currentShard.HasNext())
+                {
+                    _shards.RemoveAt(_shardIndex);
+                }
+
+                i++;
+                _shardIndex++;
```
Comment (ljian3377) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Segment.cs:95]: If `_shards.RemoveAt(_shardIndex);` , then by `_shardIndex++` unconditionally we skipped the shard that is moved to `_shardIndex` due to the removal this round.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeedFactory.cs (hunk 14 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeedFactory.cs:74]: Nope, this is by design.  It takes some time to enable change feed for an account.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeedFactory.cs [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+            // Get year paths
+            years = await GetYearPaths(async, containerClient).ConfigureAwait(false);
+
+            // Dequeue any years that occur before start time
+            if (startTime.HasValue)
+            {
+                while (years.Count > 0
+                    && years.Peek().ToDateTimeOffset() < startTime.RoundDownToNearestYear())
+                {
+                    years.Dequeue();
```
Comment (ljian3377) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeedFactory.cs:110]: Is it better if we sort the `years[]` first accroding to the number.  "Blobs are listed in alphabetical order ", then year 996 will be listed after 2020, same for 9999 and 10000, although our code probably won't run that long for the bug to kick in.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs [hunk 20]
```diff
@@ -0,0 +1,233 @@
...
+        {
+            // We have no more segments, years, and the current segment doesn't have hext.
+            if (_segments.Count == 0 && _years.Count == 0  && !_currentSegment.HasNext())
+            {
+                return false;
+            }
+
+            DateTimeOffset end = BlobChangeFeedExtensions.MinDateTime(_lastConsumable, _endTime);
+
+            return _currentSegment.DateTime <= end;
```
Comment (ljian3377) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs:133]: why returns true when `_currentSegment.DateTime == end;`

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs:133]: Because we want to include the end hour.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeedFactory.cs (hunk 19 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeedFactory.cs:110]: I don't think we need to be worried about it.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Segment.cs (hunk 18 above)
Comment (ljian3377) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Segment.cs:95]: And a more serious issue is that we mess with the `shardIndex` (which we need to return in cursor) by removing items from `_shards`, while `chunkFilePaths` stays unchanged.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeedFactory.cs (hunk 15 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeedFactory.cs:133]: fixed.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs (hunk 17 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs:106]: changed to _currentSegment.IsFinalized

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Segment.cs (hunk 18 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Segment.cs:101]: I'm changed the logic to keep a Set<Int> of finished shards, instead of removing completed shards from _shards.

**Comment by seanmcc-msft** [comment]
@ljian3377, please let me know if you find anything else.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs (hunk 20 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs:149]: Actually this might be a bug.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs [hunk 21]
```diff
@@ -107,6 +107,11 @@ internal class ChangeFeed
                 return BlobChangeFeedEventPage.Empty();
             }
 
+            if (pageSize > Constants.ChangeFeed.DefaultPageSize)
+            {
+                pageSize = Constants.ChangeFeed.DefaultPageSize;
```
Comment (ljian3377) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs:112]: Why is this needed?

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs [hunk 22]
```diff
@@ -0,0 +1,255 @@
...
+                return false;
+            }
+
+            // We have no more segments, years, and the current segment doesn't have hext.
+            if (_segments.Count == 0 && _years.Count == 0  && !_currentSegment.HasNext())
+            {
+                return false;
+            }
+
+            DateTimeOffset end = BlobChangeFeedExtensions.MinDateTime(_lastConsumable, _endTime);
```
Comment (ljian3377) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs:147]: I think we should make the logic of `HasNext()` consistent with that in the `GetPage()`, i.e. ignore `_lastConsumable` and use `_currentSegment.Finalized`.
Or the iterator will end up returning empty change feed event page forever.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Segment.cs [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+                throw new InvalidOperationException("Segment doesn't have any more events");
+            }
+
+            int i = 0;
+            while (i < pageSize && _shards.Count > 0)
+            {
+                // If this Shard is finished, skip it.
+                if (_finishedShards.Contains(_shardIndex))
+                {
+                    _shardIndex++;
```
Comment (ljian3377) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Segment.cs:84]: `_shardIndex` is not circling back to 0, so could go out of the boundary.

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/Segment.cs:84]: Fixed.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs (hunk 21 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs:112]: We want to limit page size to 5000.  The rest API list operations return a max of 5000 items per page.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChunkFactory.cs [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+        public ChunkFactory(
+            LazyLoadingBlobStreamFactory lazyLoadingBlobStreamFactory,
+            AvroReaderFactory avroReaderFactory)
+        {
+            _lazyLoadingBlobStreamFactory = lazyLoadingBlobStreamFactory;
+            _avroReaderFactory = avroReaderFactory;
+        }
+
+        public virtual Chunk BuildChunk(
+            BlobContainerClient containerClient,
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChunkFactory.cs:23]: The client doesn't change from one `BuildChunk` invocation to another. So, this seem more like a dependency here rather than method input. I'd suggest to refactor these factories in such a way that their are created with dependencies early in the top level layer.
Corresponding refactoring in Java https://github.com/Azure/azure-sdk-for-java/pull/10839/commits/6b3fa618d06b18c678e2ec3eb6bcb3a53831d7cf

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs (hunk 22 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs:147]: Fixed.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChunkFactory.cs (hunk 24 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChunkFactory.cs:23]: Made containerClient a dependency.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs [hunk 25]
```diff
@@ -133,20 +133,23 @@ internal class ChangeFeed
...
             {
                 return false;
             }
 
-            // We have no more segments, years, and the current segment doesn't have hext.
-            if (_segments.Count == 0 && _years.Count == 0  && !_currentSegment.HasNext())
+            if (_endTime.HasValue)
             {
-                return false;
+                return _currentSegment.DateTime <= _endTime;
```
Comment (ljian3377) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs:149]: I thought we agreed on `_currentSegment.DateTime < _endTime`.
Same in the `GetPage()`.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs (hunk 21 above)
Comment (ljian3377) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs:112]: Just to be clear, we are limitting it to 5000 just to follow the convention, not because it acctually has any effect in our change feed implementation, right?

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs:112]: That's correct.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs (hunk 25 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs:149]: I need to apply this change, thanks for the heads up.

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs]: Fixed.

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs]: fixed.

**Comment by seanmcc-msft** [comment]
I'm going to merge this, we can continue iterating in the 73 base branch.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/BlobChangeFeedExtensions.cs (hunk 16 above)
Comment (ljian3377) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/BlobChangeFeedExtensions.cs:92]: I think this is a bug. We shouldn't add 1 hour to the endTime if it's exactly at hour.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeedFactory.cs [hunk 26]
```diff
@@ -0,0 +1,204 @@
...
+            Queue<string> segments = new Queue<string>();
+            ChangeFeedCursor cursor = null;
+
+            // Create cursor
+            if (continuation != null)
+            {
+                cursor = JsonSerializer.Deserialize<ChangeFeedCursor>(continuation);
+                ValidateCursor(_containerClient, cursor);
+                startTime = cursor.CurrentSegmentCursor.SegmentTime;
+                endTime = cursor.EndTime;
```
Comment (ljian3377) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeedFactory.cs:58]: Is it better if `endTime = minDate(cursor.EndTime, endTime);` to allow user specify a new endTime?

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeedFactory.cs:58]: Currently, I don't provide an overload for a user to provide both a cursor and end date.  I think it's conceptually simpler to have separate overloads for cursor and start/end time.

File: sdk/storage/Azure.Storage.Blobs.ChangeFeed/README.md [hunk 27]
```diff
@@ -0,0 +1,148 @@
...
+the right to, and actually do, grant us the rights to use your contribution. For
+details, visit [cla.microsoft.com][cla].
+
+This project has adopted the [Microsoft Open Source Code of Conduct][coc].
+For more information see the [Code of Conduct FAQ][coc_faq]
+or contact [opencode@microsoft.com][coc_contact] with any
+additional questions or comments.
+
+<!-- LINKS -->
+[source]: https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/storage/Azure.Storage.Blobs/srcs
```
Comment (ljian3377) [inline sdk/storage/Azure.Storage.Blobs.ChangeFeed/README.md:133]: Linked to blob rather than change feed.

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs.ChangeFeed/README.md:133]: We will change this once Change Feed merges to master.
